### ISSUE TITLE
Replica Function support (Lambda@Edge)

### ISSUE THREAD

**Issue by iDVB** [blackflux/lambda-monitor#364]
I'm still in the process of testing out what exactly the issue might be but I have a good hunch that the service does not know how to deal with edge function logs. 

Likely this is because they're located in the region that the end user was closest to. (edge location)
This means that when an end user makes a request near Oregon.... then the logs for that lambda will end up in the Oregon region. 

Additionally, the log prefix is different as well in that its:
`/aws/lambda/us-east-1.demo-stack-proxy-dev-originRequest`
vs the traditional...
`/aws/lambda/demo-stack-proxy-dev-originRequest`

I thought I might be able to shim the issue by running around and putting `MONITORED=0` on all the replica functions... but as it appears (from the console anyway) you can't tag replica functions. 

I'm not sure how you can even programmatically detect if they are replica functions.... looking into that now.

**Comment by simlu** [comment]
So I take it deploying this monitor to all regions that could have edge function executions doesn't work (yet)? Targeting region functions log groups shouldn't be hard.

Why do you want to disable monitoring for those functions? Is it more broken than simply not working?

What issues are you seeing? As I said this service works great for us, but we are not currently using edge lambda function.

**Comment by iDVB** [comment]
No, as is, it seems to be fully broken as I think the Dummy Custom Resources that initially calls those subscribe functions are failing and causing the deploy to fail and roll back. This is maybe to do with it trying to locate a replica's associated log group and failing? 

Deploying to all regions might work but is a bit overkill in my mind, especially since the regions that are “likely to have edge functions” is all of them. That’s how they work, they get replicated to all edge locations and sit there ready anytime an end user visits from that area. So it would be impossible to predict.

I'm going to dive into the code a bit more and see if I can see exactly what the issue might be. But let me know if you have any ideas. I'm very happy to test them.

**Comment by simlu** [comment]
I'd love to see what the cloudformation template / resource lambda function logs look like (you need to be quick to find them before everything is torn down).

Also I'm sure @macmcirish has some input here in what would be interesting to look at!

**Comment by iDVB** [comment]
AFK, for 1hr then will look for ya. 
Also, I think I can alter these refs to logs and have it check for replica log groups across regions. 
https://github.com/blackflux/lambda-monitor/blob/04e8d5ec00c5085388f8223d398b7acd5bfe6966/src/logic/util/lambda.js#L28

**Comment by MacMcIrish** [comment]
@iDVB Failing on the post deploy hook via a dummy resource is a frustrating issue that I plan to make much easier to understand :) up until now, I was the only one suffering from this, so I'm sorry to put you through my pain. Fortunately this is only an issue on the _very first deploy_ of the stack, as when the initial deploy fails the teardown of the stack also deletes the log group.

To examine the log group before the teardown happens:
1. Open the [Cloudwatch Management Console](https://us-west-2.console.aws.amazon.com/cloudwatch/home?region=us-west-2#logs:) in the targeted region
2. Deploy the stack
3. As the stack is deploying, refresh the page until `/aws/lambda/lambda-monitor-${YOUR_STAGE_HERE}-subscribe` and `/aws/lambda/lambda-monitor-${YOUR_STAGE_HERE}-set-retention` appear. Open the log groups immediately.
4. When the logs are attached to this log group, you can find the error output on _why_ the custom resource failed to deploy.

Once you have this info, it'll be a lot easier to track down the reason that this is failing to stabilise on deploy. Looking forward to hearing back from you!

**Comment by iDVB** [comment]
Thanks! Actually, it seems to successfully send the error to Rollbar... so I CAN see that its due to 
> ResourceNotFoundException: The specified log group does not exist.

which led me to hypothesize about it not being able to fine the replica log groups.

Testing your steps to catch the logs though...

**Comment by iDVB** [comment]
Looks like I can see pretty much the same thing in the Rollbar logs

**Comment by iDVB** [comment]
@MacMcIrish @simlu  Any reason `lambda.js` isn't using `AWS.REQUEST.promise().then()` rather then creating your own promises? https://aws.amazon.com/blogs/developer/support-for-promises-in-the-sdk/

**Comment by simlu** [comment]
@iDVB Not really. That should probably be changed if the tests just continue working. I've already done that in other places last night.

**Comment by iDVB** [comment]
Working on a PR for this now...
https://github.com/blackflux/lambda-monitor/pull/371

**Comment by simlu** [comment]
Task outline:

(1)
[CloudTrail](https://github.com/blackflux/lambda-monitor/blob/master/serverless.core.yml#L191) needs to be getting [events from all regions](https://aws.amazon.com/blogs/aws/aws-cloudtrail-update-turn-on-in-all-regions-use-multiple-trails/)
This is necessary so we can do actions based on edge lambda function log group creation.

(2) 
The [subscribe](https://github.com/blackflux/lambda-monitor/blob/master/src/logic/subscribe.js) and [set-retention](https://github.com/blackflux/lambda-monitor/blob/master/src/logic/set-retention.js) need to be updated to handle ["CloudTrail log group events" and "non CloudTrail log groups events" differently](https://github.com/blackflux/lambda-monitor/blob/master/serverless.core.yml#L67). For CloudTrail events only the log group that triggered the invocation should be handled. Otherwise we need to do an "expensive" check all call.
This is necessary so we do not overload the system with requests, otherwise this could get expensive and we will run into throttling.

(3) 
Handle edge log group subscription correctly (when an edge log group gets created we need to check the master function for tags etc). This should be relatively easy, we just need to check if this is an edge log group and check the master function for tags.

--------------------------

I would start by figuring out what the lambda event for the CloudTrail log group creation event looks like (i.e. how much meta information is available).

**Comment by iDVB** [comment]
Going to test to see what we can see in the CT event by letting it trigger a function with 
`console.log(JSON.stringify(event, null, 2))` in it.
